### PR TITLE
Remove deprecated modules

### PR DESCRIPTION
**Pull request by vy**
This PR removes the following features/modules:

- `GelfLayout`
- `log4j-cassandra`
- `log4j-couchdb`
- `log4j-jeromq`
- `log4j-kafka`
- `log4j-layout-jackson-json`
- `log4j-layout-jackson-yaml`
- `log4j-mongodb3`
- `log4j-spring-boot`
- `log4j-spring-cloud-config`

and addresses following tickets:

- #1951 
- #1969

### Review aid

To ease the review process, I suggest the following approach:
1. check out the branch locally
3. diff it against `main` (grouping by directory helps)
4. only focus on changes that are outside the deleted features/modules

### CODE REVIEW

**Review by ppkarwasz — APPROVED** [review]
Looks good to me, apart some small remarks.

File: log4j-core-its/src/test/java/org/apache/logging/log4j/core/appender/SocketAppenderTest.java [hunk 1]
```diff
@@ -37,17 +41,13 @@
 import org.apache.logging.log4j.Level;
 import org.apache.logging.log4j.LoggingException;
 import org.apache.logging.log4j.ThreadContext;
-import org.apache.logging.log4j.core.Appender;
-import org.apache.logging.log4j.core.LogEvent;
-import org.apache.logging.log4j.core.Logger;
-import org.apache.logging.log4j.core.LoggerContext;
-import org.apache.logging.log4j.core.impl.Log4jLogEvent;
+import org.apache.logging.log4j.core.*;
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-core-its/src/test/java/org/apache/logging/log4j/core/appender/SocketAppenderTest.java:44]: Agreed. Disabled in my IDE. But let's add the plugin and work it out in a separate PR.

**Comment by ppkarwasz** [comment]
I also think that the reparenting of `log4j-spring-cloud-config-client` (commit 0121847af4fbba60a84eb52bc27e49a781cd4e71) should be backported to 2.x. Do you mind doing the cherry-pick after merging this PR?

File: log4j-distribution/pom.xml [hunk 2]
```diff
@@ -63,11 +63,6 @@
         <artifactId>snappy-java</artifactId>
         <version>${snappy.version}</version>
       </dependency>
-      <dependency>
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-distribution/pom.xml:66]: I will, but not in this PR, which is already big enough. We need to spend months anyway to sync `main` with `2.x`. There I will address this issue.

File: log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/categories/Appenders.java [hunk 3]
```diff
@@ -24,13 +24,5 @@ interface AsyncConversant {}
...
 
-    interface Cassandra {}
-
-    interface CouchDb {}
-
-    interface Kafka {}
-
-    interface MongoDb {}
-
     interface ZeroMq {}
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/categories/Appenders.java]: Fixed in af765be3165c2f34cb02c6cef4e0c24e8042264a.

File: log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java [hunk 4]
```diff
@@ -249,6 +250,11 @@ public Map<String, String> getContentFormat() {
         return CONTENT_FORMAT;
     }
 
+    public static JsonTemplateLayout createDefaultLayout() {
+        final DefaultConfiguration configuration = new DefaultConfiguration();
+        return JsonTemplateLayout.newBuilder().setConfiguration(configuration).build();
+    }
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java:256]: `Configuration` is not only used for interpolation, but accessing to the recycler factory and dynamically injecting beans.

File: log4j-core-its/src/test/java/org/apache/logging/log4j/core/appender/SocketAppenderTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core-its/src/test/java/org/apache/logging/log4j/core/appender/SocketAppenderTest.java:44]: I moved this to issue apache/logging-parent#63

File: log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java:256]: Ok, so the configuration can not be `null`.

However if I am not mistaken, the current callers of `JsonTemplateLayout#createDefaultLayout` are tests. Could you move the `new DefaultConfiguration()` into the code of the callers and pass it to this method via a parameter.

Cf. #756 for a similar problem in `PatternLayout`.

[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java:256]: I gave more details in #1551.

[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/JsonTemplateLayout.java]: Fixed in d86d82bf591a787505e7457e10d805c4482e6143.

**Comment by vy** [comment]
> I also think that the reparenting of `log4j-spring-cloud-config-client` (commit [0121847](https://github.com/apache/logging-log4j2/commit/0121847af4fbba60a84eb52bc27e49a781cd4e71)) should be backported to 2.x. Do you mind doing the cherry-pick after merging this PR?

@ppkarwasz, implemented this in 3b94a483cb78d17055713ea6facbea53703ab492.

**Pull request merged**